SUIT: Prepared manifests

chrysn opened the pull request:
### Contribution description

While SUIT can generally be used already with manifests that are "dropped into memory" (by any mechanism), the convenient SUIT worker thread mechanism so far could not be used with it.

This adds the suit_worker_try_prepare / suit_worker_trigger_prepared
pair for using the SUIT worker, and breaks suit_handle_manifest_buf out
of suit_handle_url (where the latter now calls the former).

#### By-catch

As part of factoring out reaping of the zombie worker thread, a locking
error that deadlocks the SUIT worker in case the race between the mutex
being unlocked and the thread being reaped hits the necessary handling
code is fixed (and mutex_unlock is called in the error path).

### Testing procedure

SUIT tests should pass.

https://github.com/RIOT-OS/RIOT/pull/19659 provides a demo of how the new API is used.

### Issues/PRs references

I think that the currently employed mechanism of having a resource to which a URL with the manifest gets posted is contrary to the design goals of SUIT -- the signed manifest should be what justifies the device to spend resources (eg. get data from a server), not a URI that is just *not* signed.

This PR makes it easier to implement a resource to which the manifest can be POSTed, rather than a CoAP URI that represents the manifest, as is proposed (but not mature) in 
https://github.com/RIOT-OS/RIOT/pull/19659.

[edit: Adjusted to reflect decisions made during review]

chrysn commented:
I'm not fully happy with the write lock situation -- but the low-level functions are already a bunch of mutually exclusive user-needs-to-coordinate functions. Maybe I'll still replace saul_worker_is_idle with an acquire/release pair that's needed around writes to the manifest buffer.

chrysn commented:
... and updated accordingly. While the low-level functions are wild as they were before, the worker..prepare family is now an acquire/commit pair (with the commit doubling as a panic hatch), which is easier to understand and use.

chrysn commented on sys/include/suit/transport/worker.h:
```diff
@@ -27,10 +27,22 @@
+ * This is used by the SUIT worker thread, which populates it with data when
+ * handling manifest URLs. It may only be written to in preparation of a
+ * @ref suit_handle_prepared call. */
+extern uint8_t suit_manifest_buf[SUIT_MANIFEST_BUFSIZE];
```
SGTM, especially as it allows the backend to use dynamic allocation or some other mutually exclusive operation, but

* I'd make the size user-determined (extra argument for the manifest length) -- as long as it's buffer backed, requesting too much will give an error, but it also allows more dynamic management.
* The typical way to release the mutex is to hand it over to the worker for processing -- I'm not sure we need a dedicated release function. (If we do acqire-release-startworker, then another part of the code that starts the worker a different way could overwrite the manifest in the mean time).

chrysn replied:
Updated: There's now a commit that makes the suit_worker_prepare / suit_worker_trigger_prepared an allocate-use pair (that's still implemented as a lock/release pair, but the API doesn't show that).

The manifest buffer is now static again; that's a separate commit on top of that. I don't think I can easily move the "make it public" and "make it private again" commits close enough to cancel out, so I'd either leave that as a temporary state on the history, or squash the commits all into one, or if you prefer to have the history do some history cleanup (which would likely lose some other commits but still be a bit step-wise).

chrysn commented:
CI tests briefly disabled; I'd like to rebase it onto https://github.com/RIOT-OS/RIOT/pull/19643 later to ease testing.

benpicco commented on sys/include/suit/transport/worker.h:
```diff
@@ -39,6 +39,42 @@ extern "C" {
+ * possible without priority inversion tricks -->, and is also what typical use
+ * cases need.
+ */
+uint8_t *suit_worker_prepare(size_t size);
```
If the user of this function is about to receive the manifest, they don't know the size yet - so what are they supposed to put here? 

`SUIT_MANIFEST_BUFSIZE` is not even public, so they can only guess.

How about `size_t suit_worker_prepare(void **buf)` - then it's clear we can read up to [return value] bytes into the provided buffer.

chrysn replied:
There *is* the issue that "can't give you that right now" is indistinguishable from "can't give you that ever", which CoAP should represent as 5.03 Service Unavailable (try again in 60 seconds) and 4.13 Request Entity Too Large -- but I'm not sure how that'd best be expressed.

Maybe in a `size_t suit_worker_prepare(size_t size, uint8_t **ptr);` -- where 0 indicates success, a positive number indicates that something of that size would work, and a negative number is -EAGAIN and tells we can't hand out anything right now? It'd be unusual error handling, but I have no better ideas. Do you have any better suggestions?

chrysn replied:
The latest update (https://github.com/RIOT-OS/RIOT/pull/19601/commits/800446b59b0c46dbbcaad836f56b119c40c2e451) now provides better error handling at the cost of two return-by-pointer arguments, and uses a `int suit_worker_prepare(uint8_t **buffer, size_t *size)` signature.

If an application can not at all predict what it will get and needs to take the maximum available, it could now attempt suit_worker_prepare with the maximum size, and try again once it learns a good size. I wouldn't recommend that, though, which is why I'm fully OK with that particular workflow being clumsy.

chrysn commented on sys/include/suit/transport/worker.h:
```diff
@@ -52,6 +88,20 @@ void suit_worker_trigger(const char *url, size_t len);
+ * @return 0 on success
+ *        <0 on error
+ */
+int suit_handle_prepared(const uint8_t *buffer, size_t size);
```
I thought it's a sensible consequence of handing out a pointer -- to take it at return.

We don't need it now, but given that there are already ideas around for using SUIT to not only handle firmware images but also side images (eg. peer firmware), it's an easy thing to do now and an API change later, should ever anything be done in parallel.

I can switch back if you prefer, but I think it's clearer API this way.

chrysn commented:
I've pushed this onto current master, as that allows running the tests again after https://github.com/RIOT-OS/RIOT/pull/19643 was fixed.

Also, there is now https://github.com/RIOT-OS/RIOT/pull/19659, which makes this testable more practically.

benpicco commented on sys/include/suit/transport/worker.h:
```diff
@@ -39,6 +39,48 @@ extern "C" {
+ *                         decreased to a size that would be acceptable.
+ *
+ * @return 0 on success
+ * @return -EAGAIN if the worker is currently busy.
```
I'm not a huge fan of an API that requires polling on the caller side.
Do you have a use-case in mind that requires this to be non-blocking?

chrysn replied:
That's the nature of anything related to the worker thread already -- and I think that's fine because:

a) We can't make this blocking without major work, because one can't block on a thread to terminate, only on the mutex it is holding, but even if that mutex dropping is the last thing that thread does, there is a possible race condition depending on the thread priorities. (It's the whole thread return value discussion).
b) When triggered from a CoAP server, blocking would block the whole CoAP thread, racking up retransmits and generally not serving the peers while this is being processed.

And most importantly, if this is blocked, it's because there is an active manifest transfer in process. Chances are the user will not even *want* this to block, but rather to see that something is afoot, and to come back after having reviewed the whole situation.

chrysn replied:
Or, from a different PoV: This is an allocation-ish situation (just that it doesn't happen from the general heap but from the 1-sized pool of possible transfers). A malloc doesn't block either.

benpicco replied:
Ok in that context it makes sense - and if the user is able to block, they can just call `suit_handle_prepared()` themselves.

chrysn commented:
Right now this (should) behave like suit_worker_trigger, just in two steps (prepare, trigger_prepared) instead of one.

I'm happy to change things around so that we get completion indication, but I'd prefer to do that separately and not while introducing a new feature, and then consistently across both prepared and by-URI manifests.

benpicco commented on sys/include/suit/transport/worker.h:
```diff
@@ -39,6 +39,48 @@ extern "C" {
+ * @param[in] manifest  Pointer to the manifest. This must be the return value
+ *                      of a previous call to @ref suit_worker_prepare, and is
+ *                      invalidated at some point during or after the call.
+ * @param[in] size      Number of bytes that have been prepared in @ref
```
```suggestion
 * @param[in] size      Number of bytes that have been prepared in
```

benpicco commented on sys/include/suit/transport/worker.h:
```diff
@@ -52,6 +94,20 @@ void suit_worker_trigger(const char *url, size_t len);
+ * @return 0 on success
+ *        <0 on error
+ */
+int suit_handle_prepared(const uint8_t *buffer, size_t size);
```
Ok I'm beginning to like this as a more use-friendly alternative to calling `suit_parse()` - it can be used independently from the 'prepared'/worker thread machinsm.

Maybe call it `suit_handle_manifest_buf()` instead?

chrysn replied:
Good point, done.

benpicco commented on sys/suit/transport/worker.c:
```diff
@@ -161,22 +174,73 @@ static void *_suit_worker_thread(void *arg)
+{
+    mutex_lock(&_worker_lock);
+    if (!_worker_reap()) {
+        return;
```
I don't quit understand why this is needed - also shouldn't we then unlock the mutex again?

chrysn replied:
That lock has always been in suit_worker_trigger before as well -- things were just split because suit_worker_trigger acquires the lock, whereas suit_worker_trigger_prepared already had it acquired in suit_worker_prepare (which, given its differing locking behavior, I've just renamed for consistency to suit_worker_try_prepare).

The worker lock is there to cover both access to the manifest buffer and to the thread. The lock is released by the thread. Therefore, the release of the lock just indicates that the thread is *almost* gone, not that it's gone for real -- that can only be made sure of by reaping the zombie state process (which is the thing we can't await for lack of thread_join functionality).

You're right on the unlocking part, that was a bug already -- it's fixed now and better documented; the complexity of it indicates that it might be a good idea to follow @kaspar030's suggestion of using a long-running thread (idempotently started on first use to retain the nice gc-sections/gc-functions properties) rather than struggling with locks around a thread stack.

benpicco reviewed (approved):
I agree that the complexity of the locking has grown pretty hairy - when the mutex is locked in `suit_worker_try_prepare()` it is then unlocked either in `suit_worker_trigger_prepared()`, `_worker_reap()` or `_suit_worker_thread()`.

That's downright scary, but I *think* this will work correctly - please squash.

benpicco commented on sys/suit/transport/worker.c:
```diff
@@ -114,19 +116,23 @@ int suit_handle_url(const char *url)
     suit_manifest_t manifest;
     memset(&manifest, 0, sizeof(manifest));
 
     manifest.urlbuf = _url;
     manifest.urlbuf_len = CONFIG_SOCK_URLPATH_MAXLEN;
```
Do we also need locking for the `_url` buffer?

aabadie commented on sys/include/suit/transport/worker.h:
```diff
@@ -39,19 +39,87 @@ extern "C" {
 void suit_worker_trigger(const char *url, size_t len);
 
+/**
+ * @brief   Trigger a SUIT udate via a worker thread
```
```suggestion
 * @brief   Trigger a SUIT update via a worker thread
```

chrysn replied:
Thanks, applied at several places.

chrysn commented:
Changes after approval were limited to typo corrections (35d53324655e8d2d14ac33b0a54eb11a17379e6e), a missing semicolon and missing rename (66baee6c17365cb679ea349ecdf2ebd437bed47c), and the still-independent-because-its-a-standalone-thing include fix of c0473313cd1ffc22c21aa3817cb4d3eb5d229cb1 (which is what broke the tests as in those the C file's own header that was always formally missing but not practically relevant became relevant during these changes, and the test's setup didn't do the same includes that my code did that made it not matter).

chrysn commented:
Fixed a final issue where an argument was only used in side an assert (so building with NDEBUG as in DEVELHELP=0 didn't use the variable) in 8d6a375537e0da0ea7d5c81bda57baf0f5057cd6, and squashed.

chrysn commented:
bors merge
